README.md
# Ember-spin-spinner

Simple Ember spinner component based on [spin.js](http://fgnass.github.io/spin.js/).

[![Build Status](https://travis-ci.org/RSSchermer/ember-spin-spinner.svg?branch=master)](https://travis-ci.org/RSSchermer/ember-spin-spinner)

## Installation

```
ember install ember-spin-spinner
```

## Usage

Example:

```hbs
<!-- somepage.hbs -->

<div class='loading-container'>
  {{spin-spinner lines=13 length=20 width=10}}
</div>
```

``` css
/* somepage.css */

.loading-container {

  /* The spinner is absolutely positioned (in the center by default),
   * so make sure to set the containing element's position to relative.
   */
  position: relative;
}
```
The spinner is absolutely positioned (in the center by default), so make sure to set the containing element's position
to relative. You may also want to give the containing element a minimum height.

The following properties can be set to customize a particular spinner:

* `lines` (default: `13`): the number of lines to draw.
* `length` (default: `20`): the length of each line.
* `width` (default: `10`): the line thickness.
* `radius` (default: `30`): the radius of the inner circle.
* `scale` (default: `1`): scales the overall size of the spinner.
* `corners` (default: `1`): corner roundness (0..1).
* `rotate` (default: `0`): the rotation offset.
* `direction` (default: `1`): 1 for clockwise, -1 for counterclockwise.
* `color` (default: `#333`): #rgb or #rrggbb or array of colors.
* `speed` (default: `1`): rounds per second.
* `trail` (default: `60`): afterglow percentage.
* `zIndex` (default: `2e9`): the z-index
* `shadow` (default: `false`): whether to render a shadow.
* `hwaccel` (default: `false`): whether to use hardware acceleration.
* `top` (default: `'50%'`): top position relative to parent.
* `left` (default: `'50%'`): left position relative to parent.

If you want to set different defaults for all spinners in your application, extend the component and override the
defaults with your own:

```javascript
// app/components/spin-spinner.js
import SpinSpinnerComponent from 'ember-spin-spinner/components/spin-spinner';

export default SpinSpinnerComponent.extend({
  length: 6,

  width: 2
});
```


CHANGELOG.md
# Ember-spin-spinner Change Log

## 0.4.0

Upgrade to ember-cli 3.5.0 and spin.js 4.0.0, eliminate bower dependency.

BREAKING: spinner color will no longer default to the parent elements text color, but will instead default to `#333`.

## 0.3.0

Adds the `scale` property, which can be used to scale the overall size of the spinner (thanks @mnowik!).

## 0.2.0

Upgraded to ember-cli 0.2.2 and Ember 1.11. Addon no longer relies on Ember prototype extensions.


index.js
'use strict';

const path = require('path');
const resolve = require('resolve');
const Rollup = require('broccoli-rollup');
const Funnel = require('broccoli-funnel');
const mergeTrees = require('broccoli-merge-trees');
const transformer = require('ember-cli-es6-transform');

module.exports = {
  name: require('./package').name,

  included(app) {
    // this._super.included.apply(app, arguments);
    this._super.included.apply(this, arguments);
    this.app = this._findHost();

    app.import('vendor/spin.js');
  },

  // Code borrowed from https://github.com/kiwiupover/ember-cli-spinjs
  
  treeForVendor(tree) {
    const spinJsPath = path.join(resolve.sync('spin.js'), '..');

    let allTrees = [];

    let rollupTree = new Rollup(spinJsPath, {
      rollup: {
        input: 'spin.js',
        output: {
          file: 'spin.js',
          format: 'es'
        },
        onwarn: function(warning) {

          // Suppress known error message caused by TypeScript compiled code with Rollup
          // https://github.com/rollup/rollup/wiki/Troubleshooting#this-is-undefined
          if (warning.code === 'THIS_IS_UNDEFINED') {
            return;
          }

          // eslint-disable-next-line no-console
          console.log("Rollup warning: ", warning.message);
        },
      }
    });

    const babel = this.app.project.findAddonByName('ember-cli-babel');
    const babelOptions = babel.buildBabelOptions();
    const es6Tree = transformer.es6Transform(rollupTree, babelOptions);

    allTrees.push(es6Tree);

    if (tree) {
      allTrees.push(tree);
    }

    return mergeTrees(allTrees);
  },

  treeForAddonStyles: function (tree) {
    const spinJsPath = path.join(resolve.sync('spin.js'), '..');

    let spinJsCSSTree = new Funnel(spinJsPath, {
      include: ['spin.css']
    });

    let allCSSTrees = [spinJsCSSTree];

    if (tree) {
      allCSSTrees.push(tree);
    }

    return mergeTrees(allCSSTrees);
  }
};


config/ember-try.js
/* eslint-env node */
module.exports = {
  scenarios: [
    {
      name: 'ember-1.13',
      bower: {
        dependencies: {
          'ember': 'components/ember#1.13'
        },
        resolutions: {
          'ember': '1.13'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-lts-2.8',
      bower: {
        dependencies: {
          'ember': 'components/ember#lts-2-8'
        },
        resolutions: {
          'ember': 'lts-2-8'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-lts-2.12',
      npm: {
        devDependencies: {
          'ember-source': '~2.12.0'
        }
      }
    },
    {
      name: 'ember-release',
      bower: {
        dependencies: {
          'ember': 'components/ember#release'
        },
        resolutions: {
          'ember': 'release'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-beta',
      bower: {
        dependencies: {
          'ember': 'components/ember#beta'
        },
        resolutions: {
          'ember': 'beta'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-canary',
      bower: {
        dependencies: {
          'ember': 'components/ember#canary'
        },
        resolutions: {
          'ember': 'canary'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-default',
      npm: {
        devDependencies: {}
      }
    }
  ]
};


addon/components/spin-spinner.js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { htmlSafe } from '@ember/string';
import { Spinner } from 'spin';

export default Component.extend({
  classNames: ['spin-spinner'],

  attributeBindings: ['style'],

  lines: 13,

  length: 20,

  width: 10,

  radius: 30,

  scale: 1,

  corners: 1,

  rotate: 0,

  direction: 1,

  speed: 1,

  trail: 60,

  zIndex: 2e9,

  shadow: false,

  hwaccel: false,

  color: '#333',

  left: '50%',

  top: '50%',

  style: computed('left', 'top', function () {
    return htmlSafe('position: absolute;' +
      'width: 0;' +
      'height: 0;' +
      `left: ${this.get('left')};` +
      `top: ${this.get('top')};`);
  }),

  didInsertElement() {
    let spinner = new Spinner({
      lines: this.get('lines'),
      length: this.get('length'),
      width: this.get('width'),
      radius: this.get('radius'),
      scale: this.get('scale'),
      corners: this.get('corners'),
      rotate: this.get('rotate'),
      direction: this.get('direction'),
      speed: this.get('speed'),
      trail: this.get('trail'),
      zIndex: this.get('zIndex'),
      shadow: this.get('shadow'),
      hwaccel: this.get('hwaccel'),
      left: 'auto',
      top: 'auto',
      color: this.get('color')
    }).spin(this.element);
    
    this.set('spinner', spinner);
  },

  willDestroyElement() {
    this.get('spinner').stop();
  }
});
